src/main/kotlin/org/rust/ide/highlight/RsHighlightExitPointsHandlerFactory.kt

```diff
@@ -0,0 +1,105 @@
+            return RsHighlightExitPointsHandler(editor, file, target)
+        }
+        if (elementType == IDENTIFIER) {
+            val parent = target.parent
+            if (parent is RsMacroCall && (parent.referenceName == "panic" || parent.referenceName == "try")) {
+                return RsHighlightExitPointsHandler(editor, file, target)
+            }
+        }
+        if (testForExprBeforeBlock(target)) {
+            return RsHighlightExitPointsHandler(editor, file, target)
+        }
+        return null
+    }
+
+    private fun testForExprBeforeBlock(target: PsiElement): Boolean {
+        var context: PsiElement? = target
+        while (context != null) {
+            val parent = context.parent
+            if (context is RsExpr && parent is RsBlock) {
+                return true
+            }
+            context = parent
+        }
+        return false
+    }
+
+    private class RsHighlightExitPointsHandler(editor: Editor, file: PsiFile, var target: PsiElement) : HighlightUsagesHandlerBase<PsiElement>(editor, file) {
+        override fun getTargets() = listOf(target)
+
+        override fun selectTargets(targets: MutableList<PsiElement>?, selectionConsumer: Consumer<MutableList<PsiElement>>?) {
+            selectionConsumer?.consume(targets)
+        }
+
+        private fun getFunctionOrLambda(target: PsiElement): PsiElement? {
+            var context: PsiElement? = target
+            while (context != null) {
+                if (context is RsFunction || context is RsLambdaExpr) {
+                    return context
+                }
+                context = context.parent
+            }
+            return null
+        }
+
+        override fun computeUsages(targets: MutableList<PsiElement>?) {
+            val function: PsiElement? = getFunctionOrLambda(target)
+            function?.acceptChildren(object : RsVisitor() {
+                override fun visitElement(element: PsiElement?) {
+                    element?.acceptChildren(this)
+                }
+
+                override fun visitFunction(o: RsFunction) {}
+                override fun visitLambdaExpr(o: RsLambdaExpr) {}
+                override fun visitRetExpr(o: RsRetExpr) = addOccurrence(o)
+                override fun visitTryExpr(o: RsTryExpr) = addOccurrence(o)
+
+                override fun visitTryMacroArgument(o: RsTryMacroArgument) {
+                    val macroCall = o.parent as? RsMacroCall ?: return
+                    addOccurrence(macroCall)
+                }
+
+                override fun visitFormatMacroArgument(o: RsFormatMacroArgument) {
+                    val macroCall = o.parent as? RsMacroCall ?: return
+                    if (macroCall.referenceName == "panic") {
+                        addOccurrence(macroCall)
+                    }
+                }
+
+                override fun visitExpr(o: RsExpr) {
+                    if (o.parent is RsBlock) {
+                        addOccurrence(o)
```

Comment:
If I understand correctly, you use this block to highlight the last expression in the function block. Unfortunately, it brings false highlights:

```rust
fn foo() {
    let a = { 10 };  // <-- shouldn't be highlighted
    loop {
        break  // <-- shouldn't be highlighted
    }
    return
    // ^ Caret
}
```

In general, highlighting implicit return looks nice in many functions. But if the expression is large, you might not want to have it highlighted. For instance, this is a quite common pattern with a long `match`-expression:

```rust
fn foo(a: u32) -> u32 {
    match a {
        0: 0,
        1..7: 1,
        8...21: 2,
        // many-many other arms
    }
} 
```

Maybe we can find a more intelligent way to handle such expressions? Maybe we shouldn't highlight them at all? Or only highlight "header"s of long expressions (words `match`, `if`, `loop`, function name, etc)?

Any ideas or examples from other languages that allow omit `return`s?